File: leet_DP_746_min_cost.py
#* ******************************************** *#
#*                                              *#
#*              \\             /`/``            *#
#*              ~\o o_       0 0\               *#
#*              / \__)      (u  ); _  _         *#
#*       / \/ \/  /           \  \/ \/ \        *#
#*      /(   . . )            (         )\      *#
#*     /  \_____/              \_______/  \     *#
#*         []  []               [[] [[]    *.   *#
#*         []] []]              [[] [[]         *#
#*                                              *#
#* ************************************ nuo *** *#

class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        
        n = len(cost)
        
        dp = [0] * (n + 1)
        # dp = defaultdict(lambda: 0) # defaultdict is faster
        
        i = 2
        while i <= n:
            L = dp[i - 2] + cost[i - 2];
            R = dp[i - 1] + cost[i - 1];
            dp[i] = L if L < R else R
            i += 1
        # print(dp)
        return dp[n]

File: leet_DP_746_min_cost.java
class Solution {
    public int minCostClimbingStairs(int[] cost)
    {
        int     len, p1, p2, i;
        int[]   dp;

        len = cost.length;
        dp = new int[len + 1];
        dp[0] = 0;
        dp[1] = 0;
        i = 1;
        while (++i < len + 1)
        {
            p1 = cost[i - 1] + dp[i - 1];
            p2 = cost[i - 2] + dp[i - 2];
            dp[i] = p1 < p2 ? p1 : p2;
        }
        return dp[len];
    }
}
